### PR TITLE
fix: binder runtime issues

### PR DESCRIPTION
**Pull request by ppruthi**
## Related Issues/PRs

# A few notebooks have runtime failures while running on [mybinder.org](mybinder.org). This PR fixes those issues.

## What changes are proposed in this pull request?

Modifying jupyter notebooks in `$REPO_ROOT/notebooks/features` to run on mybinder.org as well as our test infrastrucuture.

## How is this patch tested?

Our CI infra + manually running notebooks on [mybinder.org](mybinder.org)

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change any dependencies?

- [ ] No. You can skip this section.
- [ ] Yes. Make sure the dependencies are resolved correctly, and list changes here.

## Does this PR add a new feature? If so, have you added samples on website?

- [ ] No. You can skip this section.
- [ ] Yes. Make sure you have added samples following below steps.

1. Find the corresponding markdown file for your new feature in `website/docs/documentation` folder.
   Make sure you choose the correct class `estimators/transformers` and namespace.
2. Follow the pattern in markdown file and add another section for your new API, including pyspark, scala (and .NET potentially) samples.
3. Make sure the `DocTable` points to correct API link.
4. Navigate to website folder, and run `yarn run start` to make sure the website renders correctly.
5. Don't forget to add `<!--pytest-codeblocks:cont-->` before each python code blocks to enable auto-tests for python samples.
6. Make sure the `WebsiteSamplesTests` job pass in the pipeline.

AB#1914413

### CODE REVIEW

File: notebooks/features/opencv/OpenCV - Pipeline Image Transformations.ipynb [hunk 1]
```diff
@@ -130,6 +130,13 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "from synapse.ml.core.utils import current_platform\n",
+    "\n",
+    "platform = current_platform()\n",
+    "if platform == \"binder\":\n",
+    "    from IPython import get_ipython\n",
+    "\n",
+    "    !pip install matplotlib Pillow\n",
```
Comment (mhamilton723) [inline notebooks/features/opencv/OpenCV - Pipeline Image Transformations.ipynb:139]: should we put these libraries in the binder image?

File: notebooks/features/classification/Classification - Adult Census.ipynb [hunk 2]
```diff
@@ -93,12 +90,17 @@
    "cell_type": "code",
    "execution_count": null,
    "source": [
-    "if os.environ.get(\"AZURE_SERVICE\", None) != \"Microsoft.ProjectArcadia\":\n",
-    "    model.write().overwrite().save(\"dbfs:/AdultCensus.mml\")\n",
-    "else:\n",
+    "platform = current_platform()\n",
+    "if platform == \"synapse\":\n",
```
Comment (mhamilton723) [inline notebooks/features/classification/Classification - Adult Census.ipynb:94]: I know its a PITA but can we replace all of the places in the notebooks where we have this silly project arcadia thing?

Comment (ppruthi) [reply notebooks/features/classification/Classification - Adult Census.ipynb:94]: Yep - that was the plan -- wanted to see if this works fine and does not regress on E2E tests

File: core/src/main/python/synapse/ml/core/utils/platform.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        return "binder"
+    else:
+        return "unknown"
+
+
+def bootstrapSpark() -> SparkSession:
+    if current_platform() == "synapse":
+        return SparkSession.builder.getOrCreate()
+    else:
+        return SparkSession.getActiveSession()
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/platform.py:22]: i dont think this API is necessary as we can just call SparkSession.builder.getOrCreate() in pretty much all contexts. The former will be the same as getActiveSession if theres an active session

File: core/src/main/python/synapse/ml/core/utils/platform.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+def current_platform():
+    if os.environ.get("AZURE_SERVICE", None) == "Microsoft.ProjectArcadia":
+        return "synapse"
+    elif "dbfs" in os.listdir("/"):
+        return "databricks"
+    elif os.environ.get("BINDER_LAUNCH_HOST", None) is not None:
+        return "binder"
+    else:
+        return "unknown"
+
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/platform.py:16]: We might also want to add a findSecret API that will run through the gambit of looking in keyvaults, using databricks methods, and checking env vars before yelling and asking someone to replace that line with their key if they cant find anything

File: core/src/main/python/synapse/ml/core/utils/Platform.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    return False
+
+
+def RunningOnDatabricks():
+    if CurrentPlatform() is PLATFORM_DATABRICKS:
+        return True
+    return False
+
+
+def PrintKeyWarning():
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/Platform.py:41]: nit: use snake_case here

File: core/src/main/python/synapse/ml/core/utils/Platform.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    if not RunningOnSynapse():
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/Platform.py:42]: nit, throw an error here and refactor this to try to find a secret in a few places

File: core/src/main/python/synapse/ml/core/utils/Platform.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        from pyspark.dbutils import DBUtils
+
+        spark = SparkSession.builder.getOrCreate()
+        dbutils = DBUtils(spark)
+        return dbutils.secrets.get(scope=SECRET_STORE, key=searchKey)
+    else:
+        raise RuntimeError(
+            "#### Please add your environment/service specific key(s) before running the notebook ####"
+        ) from None
+        return ""
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/Platform.py:53]:
```suggestion
        raise RuntimeError(f"Could not find {secret_name} in keyvault or overrides. If you are running this demo and would like to manually specify your key please add the override="YOUR_KEY_HERE" to the arguments of the find_secret method")
        ) from None
        return ""
```

File: core/src/main/python/synapse/ml/core/utils/Platform.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+def running_on_binder():
+    return current_platform() is PLATFORM_BINDER
+
+
+def running_on_databricks():
+    return current_platform() is PLATFORM_DATABRICKS
+
+
+def get_platform_specific_secret(searchKey):
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/Platform.py:37]:
```suggestion
def find_secret(secret_name, keyvault=SECRET_STORE, override=None):
```

File: core/src/main/python/synapse/ml/core/utils/Platform.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        from notebookutils.mssparkutils.credentials import getSecret
+
+        return getSecret(SECRET_STORE, searchKey)
+    elif running_on_databricks():
+        from pyspark.sql import SparkSession
+        from pyspark.dbutils import DBUtils
+
+        spark = SparkSession.builder.getOrCreate()
+        dbutils = DBUtils(spark)
+        return dbutils.secrets.get(scope=SECRET_STORE, key=searchKey)
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/Platform.py:48]:
```suggestion
        return dbutils.secrets.get(scope=keyvault, key=secret_name)
```

File: core/src/main/python/synapse/ml/core/utils/Platform.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+def running_on_databricks():
+    return current_platform() is PLATFORM_DATABRICKS
+
+
+def get_platform_specific_secret(searchKey):
+    if running_on_synapse():
+        from notebookutils.mssparkutils.credentials import getSecret
+
+        return getSecret(SECRET_STORE, searchKey)
```
Comment (mhamilton723) [inline core/src/main/python/synapse/ml/core/utils/Platform.py:41]:
```suggestion
        return getSecret(keyvault, secret_name)
```

**Pull request merged**